File: Programmers/[72415] 카드 짝 맞추기/KingDonggyu/72415.js
function solution(board, r, c) {
  const bfs = (removed, src, dst) => {
    const isOutOfBounds = (x, y) => {
      return x < 0 || x > 3 || y < 0 || y > 3;
    };

    const di = [[-1, 0], [1, 0], [0, -1], [0, 1]];
    const visited = Array.from(Array(4), () => Array(4).fill(false));
    const queue = [[...src, 0]];

    while (queue.length) {
      const [x, y, cnt] = queue.shift();
      if (x === dst[0] && y === dst[1]) {
        return cnt;
      }
      // 상, 하, 좌, 우 이동
      for (const [dx, dy] of di) {
        let [nx, ny] = [x + dx, y + dy];

        if (isOutOfBounds(nx, ny)) {
          continue;
        }

        if (!visited[nx][ny]) {
          visited[nx][ny] = true;
          queue.push([nx, ny, cnt + 1]);
        }

        // [Ctrl] 키 누르기
        // - 이미 한칸 이동했으므로 최대 2칸 이동 가능
        for (let i = 0; i < 2; i++) {
          // 카드가 존재
          if (!(removed & (1 << board[nx][ny]))) {
            break;
          }
          // 커서가 더이상 이동 불가
          if (isOutOfBounds(nx + dx, ny + dy)) {
            break;
          }
          [nx, ny] = [nx + dx, ny + dy];
        }

        if (!visited[nx][ny]) {
          visited[nx][ny] = true;
          queue.push([nx, ny, cnt + 1]);
        }
      }
    }

    return Infinity;
  };

  const permutate = (cnt, removed, src) => {
    // 모든 카드 제거
    if (removed === allRemoved) {
      answer = Math.min(answer, cnt);
      return;
    }

    for (const [card, pos] of cardPos.entries()) {
      // 이미 제거한 카드
      if (removed & (1 << card)) {
        continue;
      }

      const one = bfs(removed, src, pos[0]) + bfs(removed, pos[0], pos[1]) + 2;
      const two = bfs(removed, src, pos[1]) + bfs(removed, pos[1], pos[0]) + 2;

      permutate(cnt + one, removed | (1 << card), pos[1]);
      permutate(cnt + two, removed | (1 << card), pos[0]);
    }
  };

  const cardPos = new Map();
  let allRemoved = 1;
  let answer = Infinity;

  board.forEach((row, x) => {
    row.forEach((card, y) => {
      if (!card) {
        return;
      }
      if (cardPos.has(card)) {
        cardPos.get(card).push([x, y]);
        return;
      }
      cardPos.set(card, [[x, y]]);
      // 카드 값에 대한 비트 마스킹
      // - 비트 1이면 제거됨을 의미
      allRemoved |= 1 << card;
    });
  });

  permutate(0, 1, [r, c]);

  return answer;
}

File: Programmers/[72415] 카드 짝 맞추기/SUbbb/72415.java
import java.util.*;

class Card {
    // 행, 열, 이동하기 위해 필요한 조작 횟수
    int r;
    int c;
    int count;
    
    public Card(int r, int c, int count) {
        this.r = r;
        this.c = c;
        this.count = count;
    }
}

class Solution {
    static int[][] board;
    static int[][] range = {{ -1, 0 }, { 1, 0 }, { 0, -1 }, { 0, 1 }};
    
    public int solution(int[][] board, int r, int c) {
        this.board = board;
        
        return permutate(new Card(r, c, 0));
    }
    
    private static int permutate(Card src) {
        int ret = Integer.MAX_VALUE;
        
        // 1부터 6까지의 카드가 존재할 수 있다.
        for (int n = 1; n <= 6; n++) {
            // n에 해당하는 카드를 저장하는 리스트
            ArrayList<Card> cards = findNCard(n);
            
            // 선택한 카드들이 없는 경우, 패스
            if (cards.isEmpty()) continue;
            
            // 현재 위치부터 선택한 카드까지 가기 위한 거리를 계산
            int one = bfs(src, cards.get(0)) + bfs(cards.get(0), cards.get(1)) + 2;
            // one의 역순으로 진행
            int two = bfs(src, cards.get(1)) + bfs(cards.get(1), cards.get(0)) + 2;
            
            // 뒤집은 카드를 보드에서 제거
            removeCards(cards);
            
            // 재귀 호출 수행
            // one으로 수행한 카드 뒤집기 이후, 커서가 card.get(1)에 있을 때 남은 카드들에 대한 뒤집기 수행
            ret = Math.min(ret, one + permutate(cards.get(1)));
            // two로 수행한 카드 뒤집기 이후, 커서가 card.get(0)에 있을 때 남은 카드들에 대한 뒤집기 수행 
            ret = Math.min(ret, two + permutate(cards.get(0)));
            
            // 1번 카드에 대한 뒤집기 수행 후, 2번 카드에 대한 뒤집기 진행 시에는 1번 카드 뒤집기에 대한 복원이 필요하다.
            restoreCards(cards, n);
        }
        
        if (ret == Integer.MAX_VALUE) return 0;
        
        return ret;
    }
    
    private static int bfs(Card src, Card dest) {
        boolean[][] visited = new boolean[4][4];
        Queue<Card> queue = new LinkedList<>();
        queue.add(src);
        
        while(!queue.isEmpty()) {
            Card now = queue.poll();
            
            // 목적지에 도착한 경우, 조작 횟수를 반환
            if (now.r == dest.r && now.c == dest.c) return now.count;
            
            for (int i = 0; i < 4; i++) {
                int nr = now.r + range[i][0], nc = now.c + range[i][1];
                
                if (!isInBoundary(nr, nc)) continue;
                
                // 방문하지 않은 경우 조작 횟수 갱신
                if (!visited[nr][nc]) {
                    visited[nr][nc] = true;
                    queue.add(new Card(nr, nc, now.count + 1));
                }
                 
                // ctrl을 이용해 조작 가능한 경우
                for (int j = 0; j < 2; j++) {
                    // 뒤집어진 카드가 아닌 경우 종료
                    if (board[nr][nc] != 0) break;
                    // 한 번 더 해당 방향으로 진행했을 때 경계를 벗어나는 경우도 종료
                    if (!isInBoundary(nr + range[i][0], nc + range[i][1])) break;
                    
                    nr += range[i][0];
                    nc += range[i][1];
                }
                
                // nr, nc는 이제 한 번에 이동할 수 있는 위치
                if (visited[nr][nc]) continue;
                visited[nr][nc] = true;
                queue.add(new Card(nr, nc, now.count + 1));
            }
        }
        
        return Integer.MAX_VALUE;
    }
    
    private static boolean isInBoundary(int r, int c) {
        return r >= 0 && r < 4 && c >= 0 && c < 4;
    }
    
    private static ArrayList<Card> findNCard(int num) {
        ArrayList<Card> cards = new ArrayList<>();
        
        for (int i = 0; i < 4; i++)
            for (int j = 0; j < 4; j++)
                if (board[i][j] == num)
                    cards.add(new Card(i, j, 0));
        
        return cards;
    }
    
    private static void removeCards(ArrayList<Card> cards) {
        for (Card card : cards)
            board[card.r][card.c] = 0;
    }
    
    private static void restoreCards(ArrayList<Card> cards, int num) {
        for (Card card : cards)
            board[card.r][card.c] = num;
    }
}